### PR TITLE
ENH: Removing TransformType key from Transform serialization

### PR DESCRIPTION
**Pull request by PranjalSahu**
- Removed TransformType key from Transform dict.
- Added test for BSpline Transform for testing transformation output before and after serialization.
- Removed the first Composite transform from the dict array. However, there is one issue with this approach. The dictionary is a list of transforms present inside the composite transform and the name for composite transform cannot be transferred here.  We need to decide if that will be needed. 
- Changed parameter setting from loop to one function call. Construct a new itkOptimizerParameters object by converting NumPy array to a list.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

Sample Json for single Transform object:
```yaml
[
{'name': '', 
'parametersValueType': 'float64', 
'inputDimension': 2, 
'outputDimension': 2, 
'transformName': 'BSplineTransform', 
'inputSpaceName': '2D_VectorSpace', 
'outputSpaceName': '2D_VectorSpace', 
'parameters': array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  2.,  2.,  5.,  5.,  2.,  2.,  0.,  0.,  0.,  0.,  0.,  0.,
        5.,  5., 10., 10.,  5.,  5.,  0.,  0.,  0.,  0.,  0.,  0.,  2.,
        2.,  5.,  5.,  2.,  2.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.]), 
'fixedParameters': array([  6.        ,   6.        , -86.04166667, -86.04166667,
        85.41666667,  85.41666667,   1.        ,   0.        ,
         0.        ,   1.        ]), 
'numberOfParameters': 72, 
'numberOfFixedParameters': 10
}
]
```

Example of composite transform dict:
```yaml
[
{
'name': 'transform0', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'AffineTransform', 
'inputSpaceName': 'fixedSpace0', 
'outputSpaceName': 'movingSpace0', 
'parameters': array([1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.]), 
'fixedParameters': array([0., 0., 0.]), 
'numberOfParameters': 12, 
'numberOfFixedParameters': 3
}, 
{
'name': 'transform1', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'DisplacementFieldTransform', 
'inputSpaceName': 'fixedSpace1', 
'outputSpaceName': 'movingSpace1', 
'parameters': array([], dtype=float64),
 'fixedParameters': array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0.]), 
'numberOfParameters': 0, 
'numberOfFixedParameters': 18
},
{
'name': 'transform2', 
'parametersValueType': 'float64', 
'inputDimension': 3, 
'outputDimension': 3, 
'transformName': 'Rigid3DTransform', 
'inputSpaceName': 'fixedSpace2', 
'outputSpaceName': 'movingSpace2', 
'parameters': array([1., 0., 0., 0., 1., 0., 0., 0., 1., 0., 0., 0.]), 
'fixedParameters': array([0., 0., 0.]), 
'numberOfParameters': 12, 
'numberOfFixedParameters': 3
}
]****

### CODE REVIEW

**Comment by PranjalSahu** [comment]
@thewtex Do we want to give the functionality to change the **inputSpace** and **outputSpace** fields ?
For this, we will have to add string entries in the [itk::Transform](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Transform/include/itkTransform.hxx) class definition and not just in the dict_from_transform.

**Comment by thewtex** [comment]
> add string entries in the itk::Transform class definition and not just in the dict_from_transform.

@PranjalSahu yes, please

File: Modules/Core/Transform/include/itkTransform.h
Comment (jhlegarreta) [inline Modules/Core/Transform/include/itkTransform.h:124]: Please exercise the added methods in the corresponding C++ test.

**Review by jhlegarreta — DISMISSED** [review]

**Review by thewtex — COMMENTED** [review]
@PranjalSahu awesome, thank you! I like each transform name having the `name`, `inputSpace`, `outputSpace`.  We have the composite transform `inputSpace`, `outputSpace` by looking at `composite[0]['inputSpace']`, `composite[-1]['outputSpace']` -- best not to duplicate that information.

One minor change as suggested by @jhlegarreta 

```
 std::cout << transform->GetInputSpaceName() << std::endl;
```

->

```
ITK_TEST_EXPECT_EQUAL(std::string("test_inputspace"), transform->GetInputSpaceName());
```

File: Modules/Core/Transform/include/itkTransform.h
Comment (jhlegarreta) [reply Modules/Core/Transform/include/itkTransform.h:124]: Done in 5da379a. Thanks @PranjalSahu and @thewtex for the comment to ensure that the calls return what is expected.

**Review by jhlegarreta — COMMENTED** [review]
Looks good to me, but I am deferring to @thewtex the approval.

BTW @PranjalSahu only one of the commit messages has a body, and the imperative mood should be used in the subject line. See [CONTRIBUTING.md#commit-messages](https://github.com/InsightSoftwareConsortium/ITK/blob/23da3dfa79152648f3d38050ee147df47f41ce0a/CONTRIBUTING.md#commit-messages).

**Comment by PranjalSahu** [comment]
> Looks good to me, but I am deferring to @thewtex the approval.
> 
> BTW @PranjalSahu only one of the commit messages has a body, and the imperative mood should be used in the subject line of any commit. See [CONTRIBUTING.md#commit-messages](https://github.com/InsightSoftwareConsortium/ITK/blob/23da3dfa79152648f3d38050ee147df47f41ce0a/CONTRIBUTING.md#commit-messages).

I have added more messages and changed the commit subject.

**Comment by PranjalSahu** [comment]
I am doing some additional testing locally with the ICON registration transform.

**Comment by jhlegarreta** [comment]
> I have added more messages and changed the commit subject.

:100: Thanks.

**Review by thewtex — APPROVED** [review]

**Comment by PranjalSahu** [comment]
Please wait for my confirmation before merging.

**Comment by PranjalSahu** [comment]
Ok, it can be merged 👍.
Tested the output with ICON registration model also.

**Review by thewtex — CHANGES_REQUESTED** [review]
@PranjalSahu awesome!

One minor change noted inline per discussion.

**Comment by thewtex** [comment]
CC @HastingsGreer

**Review by thewtex — APPROVED** [review]

**Comment by thewtex** [comment]
:100: :1st_place_medal:

**Pull request merged**
